### ISSUE TITLE
Kubeapps integration with AKS-managed Azure Active Directory integration

### ISSUE THREAD

**Issue by abarqawi** [kubeapps/kubeapps#2118]
### Description:

we are using kubeapps with AKS-managed AAD 
we tried to follow the same  guide here 
https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md#azure-active-directory
and update the values file accordingly https://github.com/kubeapps/kubeapps/blob/master/chart/kubeapps/values.yaml#L547
but  We get a 401 Unauthorized, see attached image.

I believe this is because we are using the app registration that used to handle RBAC for the cluster 

I have also attached a screenshot of the relevant section in our values.yaml.
We are deploying it using helm like this:
helm --namespace kubeapps upgrade --install kubeapps bitnami/kubeapps --version 3.5.2 --values ./values.yaml --set "authProxy.clientSecret=${CLIENT_SECRET}" --set "authProxy.cookieSecret=$(openssl rand -base64 24)"

What we need to figure out is how to configure Kubeapps so that it can authenticate users towards the Kubernetes API server using RBAC with our AAD, in such a way that each user gets logged in with its own profile and their respective permissions in Kubeapps like they would be when using kubectl on their computers.

**Note:**
as per this guide its
https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md#azure-active-directory
 targeting the old legacy way for AKS https://docs.microsoft.com/en-us/azure/aks/azure-ad-integration-cli (legacy)
we are using the new approach here https://docs.microsoft.com/en-us/azure/aks/managed-aad
so wondering what is the correct approach or how to guide for integration?

specially in the new approach there is no Server app , or we should add it explicitly ?
https://docs.microsoft.com/en-us/azure/aks/managed-aad#create-an-aks-cluster-with-azure-ad-enabled

"AADProfile": {
    "adminGroupObjectIds": [
      "5d24****-****-****-****-****afa27aed"
    ],
    "clientAppId": null,
    "managed": true,
    "serverAppId": **null**,
    "serverAppSecret": **null**,
    "tenantId": "72f9****-****-****-****-****d011db47"
  }

![image](https://user-images.githubusercontent.com/16225699/96571035-b9d3d880-12d3-11eb-9d8c-c02017f11103.png)

### Steps to reproduce the issue:

1.Created AKS cluster with AKS-managed Azure Active Directory integration , following this URL 
https://docs.microsoft.com/en-us/azure/aks/managed-aad

2. Install the kubeapp using 

helm --namespace kubeapps upgrade --install kubeapps bitnami/kubeapps --version 3.5.2 --values ./values.yaml --set "authProxy.clientSecret=${CLIENT_SECRET}" --set "authProxy.cookieSecret=$(openssl rand -base64 24)"

3. fix the values file according to Azure AAD 

### Describe the results you received:

We get a 401 Unauthorized,

![image](https://user-images.githubusercontent.com/16225699/96571193-ec7dd100-12d3-11eb-983c-2150521542c3.png)

we expect to allow Azure AAD users to login to KubeApps and impersonate them and allow to deploy apps
is this possible?

### Version of Helm, Kubeapps and Kubernetes:

- Output of `helm version`:

```
(paste your output here)
```

- Output of `helm list -n <kubeapps-namespace> <kubeapps-release-name>`:

```
(paste your output here)
```

- Output of `kubectl version`:

```
(paste your output here)
```

**Comment by andresmgot** [comment]
Can you post the values you are using in the `values.yaml`? (I don't see them in the original comment) Maybe we can spot there a missing flag.

> we expect to allow Azure AAD users to login to KubeApps and impersonate them and allow to deploy apps
is this possible?

You don't need to do any impersonation, users will be able to login using Azure AAD and use the permissions associated to their account/group.

**Comment by lindhe** [comment]
Hi @andresmgot , appreciate that you are looking into this. For context, this is the issue we have been struggling with as of late, and  @abarqawi is the Microsoft support representative helping us to figure out the details. I think all three parties here (Microsoft, Bitnami and LKAB) need to be part of this discussion to figure this out. We are all sitting of different puzzle pieces, I believe.

So with the new "Managed AAD" feature, there seems to be no Azure App Registration (i.e., client-id and client-secret) for us to use. Either we need to get that from Azure somehow, or we need to be able to configure Kubeapps (or more specifically auth-proxy) to use another method of getting a valid token for interacting with Kubernetes RBAC.


I'm filling in our system info below. Will post the Values file soon!

## Version of Helm, Kubeapps and Kubernetes:

**Output of `helm version`:**

    version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}

**Output of `helm list -n <kubeapps-namespace> <kubeapps-release-name>`:**

    NAME    	NAMESPACE	REVISION	UPDATED                                 	STATUS  	CHART         	APP VERSION
    kubeapps	kubeapps 	1       	2020-10-15 12:53:17.513299292 +0200 CEST	deployed	kubeapps-3.5.2	v1.9.2     


**Output of `kubectl version`:**
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.7", GitCommit:"b4455102ef392bf7d594ef96b97a4caa79d729d9", GitTreeState:"clean", BuildDate:"2020-06-17T11:39:47Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.11", GitCommit:"3a3612132641768edd7f7e73d07772225817f630", GitTreeState:"clean", BuildDate:"2020-09-02T06:46:33Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by lindhe** [comment]
While writing my answer I realized my last release was an older version of Kubeapps, so I upgraded to chart version 4.0.4. Like expected, it has the same behavior. So this version of values etc. may differ every so slightly from what we showed before. But I don't think it makes a difference.

We usually use `helmfile` to deploy, but this command should be similar enough:
```console
helm --namespace kubeapps install kubeapps bitnami/kubeapps --version 4.0.4 --values ./values.yaml --set "authProxy.clientSecret=${CLIENT_SECRET}" --set "authProxy.cookieSecret=$(openssl rand -base64 24)"
```
Also worth noting that we have an nginx ingress controller and cert-manager configured in our cluster. But I'm fairly sure that's not where the issue lies here. The issue is that the "AKS Managed AAD" seems to sidestep the regular OAuth2 pattern with a client application that has a secret etc.

## `values.yaml`:
```yaml
authProxy:
  enabled: true
  # AKSAzureADServer App
  clientID: <our_client_id>
  # clientSecret: should NOT be checked in for security reasons; use --set "authProxy.clientSecret=..."
  # cookieSecret: should NOT be checked in for security reasons; use --set "authProxy.cookieSecret=..." (see README.md)
  provider: oidc
  additionalFlags:
    - --oidc-issuer-url=https://sts.windows.net/<our_tennant_id>/
    - --user-id-claim=sub

ingress:
  enabled: true
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    cert-manager.io/cluster-issuer: letsencrypt
    nginx.ingress.kubernetes.io/proxy-buffer-size: "8k"
  hostname: example.com
  # TLS secret has name {{- printf "%s-tls" .Values.ingress.hostname }} and can't be customized
  tls: true
```

**Comment by andresmgot** [comment]
I assume you already looked at this but is there any log that you can bring to try to detect the source of the issue? For example, the logs of the kube-api server should contain more info of why the request has been unauthorized. The ingress controller may also contain a bit more of information.

Also, is the `user-id-claim` needed? (just curious about that one).

**Comment by abarqawi** [comment]
@lindhe can you enable master node logs as per https://docs.microsoft.com/en-us/azure/aks/view-master-logs#enable-diagnostics-logs and repro the login ?
also i can get some logs from backend

**Comment by lindhe** [comment]
Not sure which log categories you want. `kube-apiserver` logs was active since before, and now I activated `kube-controller-manager` logs too. I reproduced the error ca 14:42 UTC today.

**Comment by andresmgot** [comment]
and what are those logs? Can you post the error you are seeing? (just to have more information)

**Comment by lindhe** [comment]
Tbh, I don't know which of those logs might have something to tell us. Have to wait for @abarqawi and see if he knows what to look for there.

**Comment by abarqawi** [comment]
yes i noticed this error in kube-api logs related to old kubectl client version 

log	pod	container	insight
failed to verify token for azure: oidc: expected audience "6dae42f8-4368-4678-94ff-3960e28e3630" got ["spn:6dae42f8-4368-4678-94ff-3960e28e3630"]

guard-86d7bbf7c4-kbtvq		kubectl-too-old

can you check those requirement ?
https://docs.microsoft.com/en-us/azure/aks/managed-aad#prerequisites

also i noticed similar issue might help https://github.com/kubeapps/kubeapps/issues/975

**Comment by lindhe** [comment]
> yes i noticed this error in kube-api logs related to old kubectl client version
> 
> log pod container insight
> failed to verify token for azure: oidc: expected audience "6dae42f8-4368-4678-94ff-3960e28e3630" got ["spn:6dae42f8-4368-4678-94ff-3960e28e3630"]
> 
> guard-86d7bbf7c4-kbtvq kubectl-too-old

I'm almost positive that that log entry has nothing to do with Kubeapps. It occurs since some of our developers are using `kubectl` 1.17. They should be able to do that, since it is explicitly within the allowed version skew of Kubernetes. See my original description in #2045 which links to [this issue](https://github.com/MicrosoftDocs/azure-docs/issues/61959) and [this issue](https://github.com/MicrosoftDocs/azure-docs/issues/62236) where I originally complained about the broken version skew. There's also some information in issue 120090424006656 in your internal support system in Microsoft where I reported this broken version skew to you.

I happen to be using `kubectl` 1.17 now, but since I'm admin I can bypass RBAC so there's no issue for me. I think the log you find is just a warning since my commands still report 1.17. But Kubernetes API server 1.17 works fine with `kubectl` 1.17 (like it is supposed to), it's only Managed AAD RBAC integration that is broken with `kubectl` < 1.18.

We have previously (in #2045) verified that Kubeapps is using API calls that are matching the server API version. We tried to use a 1.18 cluster, to trick Kubeapps to use that API version too, but that didn't help. **The issue is that we do not know what client ID and client secret to use for the OIDC ID Provider that our Kubernetes cluster is using since Managed AAD was activated.**

Issue #975 was opened in February 2019, way before Managed AAD was GA. So they are clearly (or at least ought to be) using the old way with an App Registration as IDP and thus they have a client ID and a client secret.

**Comment by absoludity** [comment]
> > yes i noticed this error in kube-api logs related to old kubectl client version
> > log pod container insight
> > failed to verify token for azure: oidc: expected audience "6dae42f8-4368-4678-94ff-3960e28e3630" got ["spn:6dae42f8-4368-4678-94ff-3960e28e3630"]
> > guard-86d7bbf7c4-kbtvq kubectl-too-old
> 
> I'm almost positive that that log entry has nothing to do with Kubeapps. It occurs since some of our developers are using `kubectl` 1.17. They should be able to do that, since it is explicitly within the allowed version skew of Kubernetes. See my original description in #2045 which links to [this issue](https://github.com/MicrosoftDocs/azure-docs/issues/61959) and [this issue](https://github.com/MicrosoftDocs/azure-docs/issues/62236) where I originally complained about the broken version skew. There's also some information in issue 120090424006656 in your internal support system in Microsoft where I reported this broken version skew to you.
> 
> I happen to be using `kubectl` 1.17 now, but since I'm admin I can bypass RBAC so there's no issue for me. I think the log you find is just a warning since my commands still report 1.17. But Kubernetes API server 1.17 works fine with `kubectl` 1.17 (like it is supposed to), it's only Managed AAD RBAC integration that is broken with `kubectl` < 1.18.

Correct. If you look at the diff between 1.17 and 1.18 for the azure auth plugin in `client-go` used by `kubectl`, you'll see the changes (I'm just limiting here to the README for brevity):
```
~/dev/3rdparty/client-go master $ git diff kubernetes-1.17.9..kubernetes-1.18.0 plugin/pkg/client/auth/azure/README.md

diff --git a/plugin/pkg/client/auth/azure/README.md b/plugin/pkg/client/auth/azure/README.md
index 21a39ae8..096f99e1 100644
--- a/plugin/pkg/client/auth/azure/README.md
+++ b/plugin/pkg/client/auth/azure/README.md
@@ -38,7 +38,13 @@ This plugin provides an integration with Azure Active Directory device flow. If
    * Replace `APISERVER_APPLICATION_ID` with the application ID of your `apiserver` application ID
    * Be sure to also (create and) select a context that uses above user
 
- 6. The access token is acquired when first `kubectl` command is executed
+6. (Optionally) the AAD token has `aud` claim with `spn:` prefix. To omit that, add following auth configuration:
+
+   ```
+     --auth-provider-arg=config-mode="1"
+   ```
+
+ 7. The access token is acquired when first `kubectl` command is executed
```
In particular, if you look at the related changes you'll see that in 1.17 the audience is always prefixed with "spn:" (causing the error in logs that you're seeing), whereas with 1.18 and onwards it is configurable as per the readme. My bet is that your KUBECONFIG file initiaized with your azure credentials will include an `config-mode: "1"` (which will be ignored by 1.17 and found and used by 1.18).

I suspect this is the reason for the note in the MS docs saying: "Important: You must use Kubectl with a minimum version of 1.18.1 or kubelogin. If you don't use the correct version, you will notice authentication issues."

> 
> We have previously (in #2045) verified that Kubeapps is using API calls that are matching the server API version. We tried to use a 1.18 cluster, to trick Kubeapps to use that API version too, but that didn't help. **The issue is that we do not know what client ID and client secret to use for the OIDC ID Provider that our Kubernetes cluster is using since Managed AAD was activated.**
> 

Agree 100%. Managed AAD is managing them for you, but they must be available somewhere. Based on the (somewhat confusing) resolution of [this issue](https://github.com/Azure/AKS/issues/1782), it sounds like the `apiserver-id` found in your kubeconfig is the application id / client id? I don't know if the secret is also there, or whether you can find this application in the azure portal to find the secret? I guess the issue here is that you're not creating this yourself so it's not clear what the id /secret is (unlike when you [create one yourself using managed aad](https://github.com/Azure/kubelogin#service-principal-login-flow-non-interactive))

**Comment by lindhe** [comment]
It seems like there is an App Registration associated with the Service Principal for the Managed AAD connection in AKS. Its client ID can be found by running `az aks show --name <cluster_name> --resource-group <cluster_rg_name> --query 'servicePrincipalProfile.clientId'`

We're working on setting up things using those credentials instead. Will get back here when we have new findings.

**Comment by absoludity** [comment]
> **Update:** Still getting the same 401 Unauthorized error, although I'm now using the Service Principal credentials and have added API permissions to the app registration. Since this includes some credentials, we have a separate mail conversation with @abarqawi alongside this GitHub thread.

It may also be worth inspecting the `id_token` and `access_token` that you're getting after authentication, and verifying if either of those work with a direct curl to the api server. To do so, update the config of oauth2-proxy, adding the additional options `--set-authorization-header=true --pass-access-token=true --set-xauthrequest` (we already set `--pass-authorization-header=true) and then reload and inspect the 401 request headers in the browser. You should find you have the `Authorization: Bearer` header with the `id_token` and the `X-Auth-Request-Access-Token` header with the access token. The reason I'm suggesting this is to verify whether *either* the `id_token` or `access_token` work with a raw curl to the api server. If we find azure with managed AAD here is expecting the access_token, it's trivial to update kubeapps' config to use that instead (see the `proxypassAccessTokenAsBearer` chart option). If they both 401 with curl as well, we have more info for @abarqawi .

**Comment by lindhe** [comment]
Disregard the latest update about the 401. I realized that I installed it on the wrong cluster that time, so 401 would have been a perfectly reasonable response.

I've been working on getting a new cluster up and running, but there's some road bumps along the way. Will report back as soon as I have had time to evaluate the latest attempt.

Currently, it seems like the SP in our cluster is too old, so it's missing the automatic app registration that should be part of the parcel.

**Comment by lindhe** [comment]
**Update:** I tried it on a new cluster, and still get the same error.

What I did this time was to setup a new cluster with managed AAD active from the start using this command:
```console
az aks create -g rg-sum-aks-lindhe-test-4 -n sum-aks-lindhe-test-4 --enable-aad --aad-admin-group-object-ids <my_ad_group_id>
```

Unlike in our old `sum-aks-2` cluster (where we activated Managed AAD months after the cluster was created), this new cluster comes with an AD App Registration associated with it. The new application has the same name as the cluster, i.e. `sum-aks-lindhe-test-4`.

I went to that app registration and created a new Client Secret. I also went ahead and setup a new authorized redirect URI and API permission to make sure the app registration has appropriate permissions.

I then used these new credentials (client id and client secret for `sum-aks-lindhe-test-4` app registration) as credentials for auth-proxy and setup Kubeapps in the new cluster. Yet, the results are the same as before.

![Screenshot from 2020-10-28 15-13-40](https://user-images.githubusercontent.com/7773090/97456061-c641dc00-1938-11eb-80d5-0f8ebded7047.png)

**Comment by absoludity** [comment]
So the IDP (AAD) is minting an `id_token` with the audience matching the client id with which oauth2-proxy is configured (as it should), but the cluster's API server does not authorize that client-id as an audience/audience with access to the cluster api server, right?

My best guess at this point is that only the principle is able to use an `id_token` to authenticate (which you don't have the cred for) but that users authenticate with the cluster using an *access* token, not an id_token (similar to GKE). Indeed, this is what I read following the managed AAD docs to https://docs.microsoft.com/en-us/azure/aks/concepts-identity#webhook-and-api-server
 
So, my recommendation to debug further is as above: set `--set-authorization-header=true --pass-access-token=true --set-xauthrequest` for the oauth2-proxy and then you can grab both the id token from the `Authorization: Bearer ...` and the access token from the `X-Auth-Request-Access-Token` and verify with curl that the id_token doesn't work, and find out if the access token *does* work. If the access token does work, it should be as simple as setting the `proxypassAccessTokenAsBearer` chart option. If it does not, then the next step would be to decode the access token (jwt.ms) and compare its contents to the decoded access token from your kubeconfig (after getting credentials the normal way and verifying kubectl works fine etc.) Hope that helps!

**Comment by weinong** [comment]
`6dae42f8-4368-4678-94ff-3960e28e3630` is the first party AAD server app used by managed AAD. You need to configure the oauth2-proxy to request the access token with `6dae42f8-4368-4678-94ff-3960e28e3630` in the audience claim.

**Comment by lindhe** [comment]
That sounds very fancy, but I have no idea what exactly it means. Let's unpack that... :)

> `6dae42f8-4368-4678-94ff-3960e28e3630` is the first party AAD server app used by managed AAD.
What's a first party AAD server app? And how did you figure that out?

> You need to configure the oauth2-proxy to request the access token with `6dae42f8-4368-4678-94ff-3960e28e3630` in the audience claim.
I like this concrete suggestion. I'd be happy to try it out, but I'm not sure how to do that. A naïve search in [the docs](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#command-line-options) gave me no clear hint on how to do that. Is audience simply the Client ID? In that case, wouldn't I need the client secret as well?

**Comment by weinong** [comment]
I'm from AKS team. :)
AKS AAD is based on OAuth2 such that the token's audience is the resource you intend to access.
In AAD v1 where you bring your own AAD apps, this is the AAD server app you provided.
In managed AAD, this is a multi-tenant AAD app that AKS manages.
The only reason that it worked with AAD v1 is the bug where we don't validate the token's audience claim. This is fixed in the managed AAD and as such you may see error in `guard` log like
```
failed to verify token for azure: oidc: expected audience "6dae42f8-4368-4678-94ff-3960e28e3630" got ["5d00fcf3-63be-4692-957c-490515e49ad7"]
```

looking at the oauth2-proxy, the `azure` provider supports specifying the [protected resource `--resource`](https://github.com/oauth2-proxy/oauth2-proxy/blob/5cba64293848866e6d09f3bd0cbbe024cf8179fb/pkg/validation/options.go#L234), and ideally you would add `--resource=6dae42f8-4368-4678-94ff-3960e28e3630`. However, it doesn't work because oauth2-proxy *always* uses the access token destined to `--resource` to query Graph api to get user's email. This will not work because the `audience` of the access token is not Graph api.

I will open an issue and look at how we can address it in oauth2-proxy. In general, there is no need to query Graph api for email. It can be configured in the claim. However, I notificed the oauth2-proxy used in kubeapps is quite old. Will the maintainer be willing to upgrade?

**Comment by absoludity** [comment]
Hi @weinong - thanks for investigating. 

> The only reason that it worked with AAD v1 is the bug where we don't validate the token's audience claim. This is fixed in the managed AAD and as such you may see error in `guard` log like

From what you'd said above, I would have thought that the only reason that it worked with AAD v1 is that the tokens audience claim *already matches* your application's ID since they both have the same value? This doc about [Azure OpenID connect](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-the-sign-in-request) explicitly has the client id as the application id:

![client-id-app-id](https://user-images.githubusercontent.com/497518/98879980-d205d480-24da-11eb-97f0-9be9706005c9.png)

and then the later [doc page about the id tokens' payload](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#payload-claims) says that the `aud` is your app's Application ID:

![payload-aud](https://user-images.githubusercontent.com/497518/98880100-1b562400-24db-11eb-9054-9fdbcc7e7e68.png)

Unless i'm missing something, this matches the normal oauth2 expecation for the `aud` claim. And this is exactly what is being returned in @lindhe 's example above using AAD2 (that is, the id token is returned by AAD with the `aud` claim matching the client id - as per the normal flow).

But from what you are saying, the request to AAD needs to specify that the intended audience is *different* from the client id in this case. This can be done in Dex by [specifying a dex-specific scope with the request](https://dexidp.io/docs/custom-scopes-claims-clients/#cross-client-trust-and-authorized-party), whereas, you're saying that the azure provider in oauth2-proxy supports this via the `--resource` option, except that it's currently broken but you'll fix it? Great, thanks. No problem at all for us to update the oauth2-proxy version once it's released in the bitnami catalog (we're currently defaulting to 6.1.1-debian-10-r12 but can update that).

**Comment by weinong** [comment]
> From what you'd said above, I would have thought that the only reason that it worked with AAD v1 is that the tokens audience claim already matches your application's ID since they both have the same value?

Right, I kinda mixed it up. In AAD v1, you can use the same app in AKS as well as in oauth2-proxy. In this scenario, the audience is same as the client. So they match. No problem.

The problem occurs when you use different app to access the managed app in managed AAD. Plus AKS AAD is based on OAuth2 rather than OIDC. So we need to send the access token destined to this managed app instead of using id token. It's like logging to a website (kubeapp/oauth2-proxy with app1) which accesses a backend web api (kube api with app2).

The changes needed in oauth2-proxy is very simple: send the access token for the protected resource to backend. yes. I'm looking into it.

**Comment by absoludity** [comment]
> The changes needed in oauth2-proxy is very simple: send the access token for the protected resource to backend. yes. I'm looking into it.

FWIW, we (Kubeapps project) don't require that in oauth2-proxy - I mentioned above that we already support an option which re-writes the Authorization header to use the access token instead of the id token (as we need this for gke). [See above](https://github.com/kubeapps/kubeapps/issues/2118#issuecomment-718286801). But great if oauth2-proxy supports it out-of-the-box, then we can simplify our chart.

**Comment by weinong** [comment]
I made some progress with fix to https://github.com/oauth2-proxy/oauth2-proxy/issues/913 and enabling proxypassAccessTokenAsBearer. it seems to be able to talk to api-server now, but it got 502 in some requests?
```
127.0.0.1:45052 - <omitted> [2020/11/12 04:36:01] localhost:3001 GET / "/api/kube/" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36" 200 4980 0.630
127.0.0.1:45062 - <omitted> [2020/11/12 04:36:02] localhost:3001 GET / "/api/tiller-deploy/v1/releases" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36" 502 559 1.017
127.0.0.1:45052 - <omitted> [2020/11/12 04:36:02] localhost:3001 GET / "/api/v1/namespaces" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36" 502 559 1.018
```

**Comment by absoludity** [comment]
Great! Not sure about the 502s without checking the env. I assume all pods (in particular, kubeops) are running without issue? Both those requests are proxied to the kubeops service, so perhaps check the logs of that pod, though with a 502, I expect it's not running, perhaps misconfigured?

**Comment by weinong** [comment]
i dont see kubeops? does it require tiller?
```
kubeapps-7f4b998b4c-6kpsw                                     2/2     Running            0          18m
kubeapps-internal-apprepository-controller-5b75dcd65d-bpg6m   1/1     Running            0          25h
kubeapps-internal-assetsvc-59b5c86799-jpmnv                   1/1     Running            1          25h
kubeapps-internal-assetsvc-59b5c86799-rgz8w                   1/1     Running            0          25h
kubeapps-internal-dashboard-98ddc4c45-979k5                   1/1     Running            0          25h
kubeapps-internal-dashboard-98ddc4c45-clrlf                   1/1     Running            0          25h
kubeapps-internal-tiller-proxy-5b84fdf7c5-28fc7               0/1     CrashLoopBackOff   294        25h
kubeapps-internal-tiller-proxy-5b84fdf7c5-9zvdq               0/1     CrashLoopBackOff   295        25h
kubeapps-mongodb-5577b5c47c-5cq9z                             1/1     Running            0          25h
```

**Comment by weinong** [comment]
Looks like I'm using chart 3.5.2... which depends on helm 2?

**Comment by weinong** [comment]
okay! it works after installing 4.0.4 instead :)

**Comment by absoludity** [comment]
Sweet! Thanks for the oauth2-proxy fix for managed AAD!

**Comment by lindhe** [comment]
This sounds promising! Thanks guys for the effort. Will attempt to reproduce it with the fix at our end as well, just to verify that we understand how to use it now.

Can we please clarify which options should be used with the modified `oauth-proxy`?

**Comment by weinong** [comment]
@lindhe this is my setting 
```
frontend:
  proxypassAccessTokenAsBearer: true
authProxy:
  enabled: true
  provider: azure
  clientID: <AAD web app>
  clientSecret: <AAD web app password>
  cookieSecret: <cookie secret>
  additionalFlags:
  - --cookie-secure=false
  - --resource=6dae42f8-4368-4678-94ff-3960e28e3630
  - --azure-tenant=001dd595-9ead-4a29-9076-ba948d58e0a0
```
In addition to this, you need your tenant admin to grant delegation permission to the server app `6dae42f8-4368-4678-94ff-3960e28e3630`. You do that by going to `App Registrations` > `API Permission` > `Add a Permission` > `API My Organization uses` > enter `Azure Kubernetes Service AAD Server` > click `Delegated Permission` and `user.read`
![image](https://user-images.githubusercontent.com/4204090/98973813-12f4ed80-24c9-11eb-8892-92d9783bd27a.png)
Once the permission is added, as your tenant admin to grant across your tenant. this is same as what you did in AAD v1.

You can test it with `docker.io/weinong/oauth2-proxy:v6.1.1-95-gedb444a` image. this is built from https://github.com/weinong/oauth2-proxy/commit/edb444a0ed4dcc7d0c74486f63e211f771b649b4

**Comment by lindhe** [comment]
@weinong 
 
> In addition to this, you need your tenant admin to grant delegation permission to the server app `6dae42f8-4368-4678-94ff-3960e28e3630`

There is no such app registration. As mentioned above, there is an SP `6dae42f8-4368-4678-94ff-3960e28e3630` but no app registration.

![Screenshot from 2020-11-13 09-09-04](https://user-images.githubusercontent.com/7773090/99044325-e0fa8e80-258f-11eb-8075-5a92a00dd607.png)

**Comment by weinong** [comment]
@lindhe I mean to create your own AAD webapp in App Registration

**Comment by weinong** [comment]
@lindhe think of it this way: user -> kubeapps/oauth2-proxy (your web app) -> AKS AAD server app.
so you need to have a webapp for oauth2-proxy to use and configure it to access to AKS managed server app.

**Comment by latiif** [comment]
Thanks for your help and input @weinong;

However I wonder what value the flag `resource` is supposed to have? Is it the `AAD web app`'s ID? in that case shouldn't the fields `clientID` and `- --resource` have the same value? 
```YAML
authProxy:
  ...
  clientID: <AAD web app ID>
  additionalFlags:
  ...
  - --resource=<AAD web app ID>
  ...
```

Or do we need to two app registrations? one client and one server?

**Comment by weinong** [comment]
clientID: <your AAD web app ID>
--resource=<the AAD resource your web app wants to access to> # this should be 6dae42f8-4368-4678-94ff-3960e28e3630

**Comment by latiif** [comment]
It works for us now. However, we are using @weinong's image since [their PR](https://github.com/oauth2-proxy/oauth2-proxy/pull/914) is still being reviewed. Which we will continue to do until it is merged.

We can go ahead and close this issue for now. 

Maybe a follow-up issue/PR can be of value to update Kubeapps guide on [Using an OAuth2/OIDC Provider with Kubeapps](https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md#azure-active-directory).

**Comment by weinong** [comment]
I will write a doc once the oauth2 proxy PR is merged and kubeapps is bumped to pick it up

**Comment by absoludity** [comment]
> I will write a doc once the oauth2 proxy PR is merged and kubeapps is bumped to pick it up

Hi @weinong . Kubeapps is now using the oauth2-proxy version with your changes. If you have a chance, can you pls update the MS Azure section at https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md#azure-active-directory ?

**Comment by fernando-almeida** [comment]
it does not look like the aforementioned docs have been updated yet.
I'm struggling to figure out with regards to having OIDC working with AAD what the resource value should be.
I've tried inferring that from @weinong comment above where it was said that the resource should be "the AAD resource your web app wants to access to" but with regards to having kubeapps working I'm still not getting it. I was hoping that I would be able to restrict this to a particular group of users and then based on those users AAD permissions the token that would be used upon authentication should directly speak to which actions they're allowed to take on the cluster.
At this point I can't even get the login screen to appear for all the attempts that I've made getting back the error:
**AADSTS500011: The resource principal named <RESOURCE_ID> was not found in the tenant named <TENANT_ID>. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.**
I'm certain that I'm setting correct client identifier and tenant identifier in the settings and it looks like this setup does not work without providing a value resource. @weinong could you please provide some additional context as to what this resource should be semantically?

**Comment by antgamdia** [comment]
Hi @fernando-almeida , we have just updated our AAD docs. Have a look at it: https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md#aks-managed-azure-active-directory

**Comment by fernando-almeida** [comment]
That documentation is much clearer and somewhat different than what was described here as it's using OIDC provider rather than azure for auth-proxy.  I'm now able to successfully get a token even though I'm still encountering another error "An error occurred: The current account does not have access to any namespaces." that I need to sort out that I believe is unrelated from this setup, which requires me to setup supporting of roles and role binding artifacts in the k8s cluster for the namespaces I want to be able to work with.
@antgamdia I very much appreciate your reply so I could get this sorted.

**Comment by antgamdia** [comment]
Não há de quê  :) I'm so happy to know it helped ! It wasn't that easy, Azure is tricky in its implementation of the flow (it's not fully oidc compliant), but fortunately, we get it sorted out.

**Comment by ojordano** [comment]
Hello @abarqawi  ;  thank for posting this and glad to see help you got.

I am trying to set up a same thing (Authenticate Kubeapps running on an AKS trough Azure) but i have 500 Internal error after login credential passed to microsoft and it tries to answer back to mydomain-url/oauth2/callback.

I have deployed kubeapps with below values and an ingress .

**helm-kubeapps-values** : 

<pre>
frontend:
  proxypassAccessTokenAsBearer: true
  
authProxy:

  enabled: true 

  provider: "oidc"
  clientID: "************"
  clientSecret: "*********"
  cookieSecret: "*******************" 

  extraFlags:
   - --oidc-issuer-url=https://login.microsoftonline.com/_tenant-id_/v2.0 
   - --scope=openid email 6dae42f8-4368-4678-94ff-3960e28e3630/user.read 
 </pre>


**ingress-kubeapps-values** : 

<pre>
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
 name: ingress-kubeapps
 annotations:
   nginx.ingress.kubernetes.io/connection-proxy-header: keep-alive
   nginx.ingress.kubernetes.io/proxy-read-timeout: "600"
   nginx.ingress.kubernetes.io/proxy-buffer-size: "32k"
spec:
  tls:
  - hosts:
    - kubeapps.domain.local
    secretName: ingress-tls 
  rules:
  - host: kubeapps.domain.local
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: helm-kubeapps
            port: 
              number: 80
  ingressClassName: nginx
 </pre>


I am able to : 
     - Have kubeapps home page with : **LOGIN VIA OIDC PROVIDER**
     -  Redirect to microsoft login page
     - Enter my allowed azure credentials to the created app on azure

But once credentials passed and it tries to do : kubeapps.domain.local/oauth2/callback?code=_long_code_ , i have : 

**500 internal, error**

**Comment by ppbaena** [comment]
Hi @ojordano, happy to help you to configure Kubeapps.

First of all, take a look to our documentation for [configuring AAD as an OIDC provider](https://github.com/vmware-tanzu/kubeapps/blob/main/site/content/docs/latest/howto/OIDC/OAuth2OIDC-azure-active-directory.md) and also for [debugging auth failures when using OIDC](https://github.com/vmware-tanzu/kubeapps/blob/main/site/content/docs/latest/howto/OIDC/OAuth2OIDC-debugging.md), so maybe this could help to identify the problem with the integration.

If you cannot find a solution to the problem with the information provided in our docs, please feel free to reopen the issue, so we can take a look and verify both the integration and the documentation provided for configuring the AAD as an OIDC.

**Comment by ppbaena** [comment]
I have just noticed you have opened #4824 and my mate @antgamdia replied in the same line. We keep the conversation in that issue.
